### ISSUE TITLE
BUG: np.str_ and np.bytes_ object have a confusing __repr__ and __str__

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#15477]
```python
>>> s = np.str_("\0\0")
>>> len(s)
2
>>> s
''  # wat
```
```python
>>> b = np.bytes_(2)
>>> len(b)
2
>>> b
b''  # wat
```

It looks to me like we should just remove our overloads of these functions, they don't make any sense.

**Comment by pv** [comment]
Is it like this to be consistent with the overall stripping of trailing NULs in S and U dtypes in arrays? It's not just string repr, but also `x = np.array([b'a\0\0\0']); assert type(x[0]) is np.bytes_; assert x[0] == b'a'` and probably other things too.

**Comment by eric-wieser** [comment]
If that's the intent, we should be stripping the nulls upon construction, not upon printing. As a more obviously broken case, consider:
```
>>> s = np.str_("ab\0")
>>> s
"ab"
>>> s[-1:]
""
```

**Comment by pv** [comment]
Indeed, `len(x[0]) == 1`, so item extraction seems to generate NUL-stripped string scalars.

**Comment by seberg** [comment]
Yeah, the NULL stripping is always an issue we have, and something we cannot get around. I think the mental model of the NULLs being stripped when the NumPy scalar is created is the best thing we can probably do. We could think about adding a warning when null padded strings/bytes are put into a NumPy array, since they will not roundtrip correctly anyway...

**Comment by eric-wieser** [comment]
So I think after removing these overloads, our options are to have the numpy subclasses:

1. Behave exactly like the builtin base class, other than having the `.dtype` attribute
1. Strip `NULL`s on construction. Note this will cause things like `len(np.str_("a\0b")[:2]) == 1`, which is perhaps surprising
1. Strip `NULL`s on construction but warn
1. Warn if trailing `NULL`s are present, but keep them

A variant of 1 would be to also change `UNICODE_setitem` to warn when nulls are stripped, which might be more user-friendly, and would catch the behavior whetther numpy subclasses are used or not.

**Comment by seberg** [comment]
I am for Option 3.

Any option should have the same result as `np.array(string, dtype="S")[()]` and `np.array(str, "S")` and `np.string_(str)` should be as similar as possible (at the moment at least, due to `PyArray_Return`).

Since it can be a bug for someone who needs the zero bytes, I think a warning makes sense. The only question is whether the warning for some reason is prohibitively noisy. I do not expect that. It could point out that `void` can sometimes be a way to spell bytes of fixed length.

**Comment by eric-wieser** [comment]
I'm leaning towards the option 4 variant. The more I think about it, 3 violates a lot of assumptions that being a subclass of `str` would imply.

Consider:
```
def replace_ext(s):
    a, b = s.split('.',1)
    return a + ".npy"
```

If passed a `np.str_` with your proposed changes, this will remove inner nulls from `"a\0.ext"`

**Comment by seberg** [comment]
True, although that code will probably give the warning (although if you have many of that you would not notice). Its true from a subclassing perspective...
It seems we either have to sacrifice consistent subclassing or consistency with arrays? Since this is NumPy scalar, which is almost an array, I think I still lean a bit towards 3. But basically there is no good solution and either 3. or 4. seem better than the current state to me.

**Comment by Zac-HD** [comment]
I'm with @seberg - personally I'd choose consistency with arrays, but either option that emits a warning is reasonable.
